xdsl/transforms/lower_affine.py

```diff
@@ -0,0 +1,144 @@
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.ir import MLContext, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.ir.affine.affine_expr import (
+    AffineBinaryOpExpr,
+    AffineBinaryOpKind,
+    AffineDimExpr,
+    AffineSymExpr,
+)
+from xdsl.ir.core import Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def affine_expr_ops(
+    expr: affine.AffineExpr,
+    dims: list[SSAValue],
+    symbols: list[SSAValue],
+) -> tuple[list[Operation], SSAValue]:
+    """
+    Returns the operations that evaluate the affine expression when given input SSA
+    values, along with the SSAValue representing the result.
+    """
+    if isinstance(expr, AffineConstantExpr):
+        constant = arith.Constant(builtin.IntegerAttr.from_index_int_value(expr.value))
+        return [constant], constant.result
+
+    if isinstance(expr, AffineDimExpr):
+        return [], dims[expr.position]
+    if isinstance(expr, AffineSymExpr):
+        return [], symbols[expr.position]
+
+    if isinstance(expr, AffineBinaryOpExpr):
+        lhs_ops, lhs_val = affine_expr_ops(expr.lhs, dims, symbols)
+        rhs_ops, rhs_val = affine_expr_ops(expr.rhs, dims, symbols)
+
+        match expr.kind:
+            case AffineBinaryOpKind.Add:
+                op = arith.Addi(lhs_val, rhs_val)
+            case AffineBinaryOpKind.Mul:
+                op = arith.Muli(lhs_val, rhs_val)
+            case AffineBinaryOpKind.Mod:
+                op = arith.RemSI(lhs_val, rhs_val)
+            case AffineBinaryOpKind.FloorDiv:
+                op = arith.FloorDivSI(lhs_val, rhs_val)
+            case AffineBinaryOpKind.CeilDiv:
+                op = arith.CeilDivSI(lhs_val, rhs_val)
+
+        return [*lhs_ops, *rhs_ops, op], op.result
+
+    raise ValueError("Unreachable")
+
+
+def insert_affine_map_ops(
+    map: affine.AffineMapAttr | None,
+    dims: list[SSAValue],
+    symbols: list[SSAValue],
+    rewriter: PatternRewriter,
+) -> list[SSAValue]:
+    """
+    Inserts operations that evaluate the affine map when given input SSA values.
+    """
+    if map is None:
+        indices = dims
+    else:
+        indices: list[SSAValue] = []
+        for expr in map.data.results:
+            ops, val = affine_expr_ops(expr, dims, [])
+            rewriter.insert_op_before_matched_op(ops)
```

Comment:
I would personally not pass the rewriter in here and rewrite, but rather return the ops+SSAvalues and then rewrite in the RewritePattern. This way, one could use this function also in a context that does not immediately rewrite (e.g. in testing). Plus, I don't quite like having the rewriting happening somewhere else than in the RewritePattern class, just from a coding style perspective, but that's just a personal preference.